configure.ac

```diff
@@ -449,6 +449,7 @@ if test "$CXXFLAGS_overridden" = "no"; then
                         [AC_LANG_SOURCE([[struct A { virtual void f(); }; struct B : A { void f() final; };]])])
   AX_CHECK_COMPILE_FLAG([-Wunreachable-code-loop-increment], [WARN_CXXFLAGS="$WARN_CXXFLAGS -Wunreachable-code-loop-increment"], [], [$CXXFLAG_WERROR])
   AX_CHECK_COMPILE_FLAG([-Wimplicit-fallthrough], [WARN_CXXFLAGS="$WARN_CXXFLAGS -Wimplicit-fallthrough"], [], [$CXXFLAG_WERROR])
+  AX_CHECK_COMPILE_FLAG([-Wdeprecated-enum-enum-conversion], [WARN_CXXFLAGS="$WARN_CXXFLAGS -Wdeprecated-enum-enum-conversion"], [], [$CXXFLAG_WERROR])
```

Comment:
Looks like his is enabled by default when targeting C++20 in GCC, and by default in Clang, so I don't think there's a need to add this explicitly.

Reply:
I presume it is not enabled by default in C++17, which is used to compile Bitcoin Core (and will be for the coming months/years)?

Reply:
It's enabled by default in Clang, which should be good enough coverage for us.
In GCC, when targetting pre-C++20, it's enabled by `-Wenum-conversion`, which is part of `-Wextra`, which we already enable. Although in that case it may only be applied to C code.

Reply:
> I presume it is not enabled by default in C++17

Correct.

Reply:
> It's enabled by default in Clang, which should be good enough coverage for us.

For me it is only enabled with c++20. Do you have steps to reproduce?

Reply:
> Do you have steps to reproduce?

No. I assumed Clang might be enabling with any standard, given they [don't specify otherwise](https://clang.llvm.org/docs/DiagnosticsReference.html#wdeprecated-enum-enum-conversion), like GCC does. However, I'm still a NACK on adding this. It's already enabled by default when it's needed, it's only available on fairly "new" compilers, so having everyone building with older ones check for availability is pointless, and I'd prefer that our configure checks don't return to being a grab-bag of checks for random flags. In any case, given the only existence of warning producing code, are a few instances in the GUI, I'd say the likelihood of any warning producing code being introduced is a also very low.

Reply:
> I'm still a NACK on adding this.

An additional minor pro for adding this line is simplicity of testing this PR by reviewers using gcc.

Reply:
> An additional minor pro for adding this line is simplicity of testing this PR by reviewers using gcc.

I don't really think we should be adding warning flags to our build system for the sake of making testing PRs easier. If anyone wants to test this change, they can do `./configure CXXFLAGS="-Wdeprecated-enum-enum-conversion"` using a compiler that supports the option.

Reply:
[Updated](https://github.com/bitcoin/bitcoin/pull/24624#issuecomment-1075072189).